core/trino-main/src/test/java/io/trino/server/TestNodeStateManager.java

```
@@ -0,0 +1,424 @@
+
+        Supplier<List<TaskInfo>> taskInfoSupplier = () -> tasks.get();
+        return new NodeStateManager(
+                sqlTasksObservable,
+                taskInfoSupplier,
+                serverConfig,
+                shutdownAction,
+                new LifeCycleManager(Collections.emptyList(), null),
+                executor);
+    }
+
+    private static class TestTaskObservable
+            implements NodeStateManager.SqlTasksObservable
+    {
+        Map<TaskId, StateChangeListener<TaskState>> tasks = new HashMap<>();
+
+        @Override
+        public void addStateChangeListener(TaskId taskId, StateChangeListener<TaskState> stateChangeListener)
+        {
+            tasks.put(taskId, stateChangeListener);
+        }
+
+        Map<TaskId, StateChangeListener<TaskState>> getTasks()
+        {
+            return tasks;
+        }
+    }
+
+    record Job(Runnable command, long time) {}
+
+    private static class TestingShutdownAction
+            implements ShutdownAction
+    {
+        private boolean shuttingDown;
+
+        @Override
+        public void onShutdown()
+        {
+            shuttingDown = true;
+        }
+
+        public boolean isShuttingDown()
+        {
+            return shuttingDown;
+        }
+    }
+
+    private static class FakeScheduledExecutorService
+            implements ScheduledExecutorService
+    {
+        private final Ticker clock;
+        private List<Job> jobs = new ArrayList<>();
+
+        public FakeScheduledExecutorService(Ticker clock)
+        {
+            this.clock = clock;
+        }
+
+        @Override
+        public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit)
+        {
+            jobs.add(new Job(command, clock.read() + unit.toNanos(delay)));
+            return null;
+        }
+
+        @Override
+        public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit)
+        {
+            return null;
```

I would throw on all unsupported methods.